### DISCUSSION TITLE
Does a TaskGroup keep references to finished tasks?

### DISCUSSION THREAD

**Discussion by Tinche** [agronholm/anyio#395] [Q&A]
I want to have a very long-lived `TaskGroup` (as part of a connection pool that lives as long as the process, essentially) and submit tasks to it. As long as these tasks finish in a timely fashion, will this cause a memory leak?

**Comment by TayfurYldz** [comment]
Finished child tasks are removed from the task group's bookkeeping; a long-lived task group does not keep every successfully completed task forever.

In the current asyncio backend, each spawned task is added to `TaskGroup._tasks`, but its done callback explicitly does:

```python
task_state.cancel_scope._tasks.remove(_task)
self._tasks.remove(task)
del _task_states[_task]
```

so the strong references used for active-task tracking are released when the task completes. The global task-state mapping is also a `WeakKeyDictionary`.

The Trio backend delegates the corresponding lifetime management to Trio's nursery.

There is one important distinction: an **exception** from a child task is retained in the task group's exception collection until the group exits, because the group must eventually raise/report those errors. So a process-long task group whose children repeatedly fail is not equivalent to one whose children finish normally; retained exception objects can also retain tracebacks and their referenced locals.

For the case you described — lots of tasks that complete successfully in a timely fashion — the task group itself isn't expected to accumulate references to all historical tasks. If you're diagnosing growth in a real pool, I'd separately check task results/closures held by your application and the failure path, rather than assuming the nursery/task-group active set is an append-only history.

I verified this against current AnyIO's asyncio backend before answering; the cleanup happens directly in the spawned task's done callback.
